src/module.cc

```diff
@@ -0,0 +1,338 @@
+  }
+}
+
+// Watchdog thread implementation for signal-triggered NodeReport
+inline void* ReportSignalThreadMain(void* unused) {
+  for (;;) {
+    uv_sem_wait(&report_semaphore);
+    if (nodereport_verbose) {
+      fprintf(stderr, "Signal %s received by nodereport module\n", node::signo_string(report_signal));
+    }
+    uv_mutex_lock(&node_isolate_mutex);
+    if (auto isolate = node_isolate) {
+      // Request interrupt callback for running JavaScript code
+      isolate->RequestInterrupt(SignalDumpInterruptCallback, &report_signal);
+      // Event loop may be idle, so also request an async callback
+      size_t signo_data = static_cast<size_t>(report_signal);
+      nodereport_trigger_async.data = reinterpret_cast<void *>(signo_data);
+      uv_async_send(&nodereport_trigger_async);
+    }
+    uv_mutex_unlock(&node_isolate_mutex);
+  }
+  return nullptr;
+}
+
+static int StartWatchdogThread(void *(*thread_main) (void* unused)) {
+  pthread_attr_t attr;
+  pthread_attr_init(&attr);
+
+  pthread_attr_setstacksize(&attr, PTHREAD_STACK_MIN);
+  pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+  sigset_t sigmask;
+  sigfillset(&sigmask);
+  pthread_sigmask(SIG_SETMASK, &sigmask, &sigmask);
+  pthread_t thread;
+  const int err = pthread_create(&thread, &attr, thread_main, nullptr);
+  pthread_sigmask(SIG_SETMASK, &sigmask, nullptr);
+  pthread_attr_destroy(&attr);
+  if (err != 0) {
+    fprintf(stderr, "nodereport: pthread_create: %s\n", strerror(err));
+    fflush(stderr);
+    return -err;
+  }
+  return 0;
+}
+#endif
+
+
+/*******************************************************************************
+ * Native module initializer function, called when the module is require'd
+ *
+ ******************************************************************************/
+void Initialize(v8::Local<v8::Object> exports) {
+  v8::Isolate* isolate = Isolate::GetCurrent();
+  node_isolate = isolate;
+
+  SetLoadTime();
+
+  const char* trigger_events = getenv("NODEREPORT_EVENTS");
+  if (trigger_events != NULL) {
+    nodereport_events = ProcessNodeReportEvents(trigger_events);
+  }
+  const char* core_dump_switch = getenv("NODEREPORT_COREDUMP");
+  if (core_dump_switch != NULL) {
+    nodereport_core = ProcessNodeReportCoreSwitch(core_dump_switch);
+  }
+  const char* trigger_signal = getenv("NODEREPORT_SIGNAL");
+  if (trigger_signal != NULL) {
+    nodereport_signal = ProcessNodeReportSignal(trigger_signal);
+  }
+  const char* report_name = getenv("NODEREPORT_FILENAME");
+  if (report_name != NULL) {
+    ProcessNodeReportFileName(report_name);
+  }
+  const char* directory_name = getenv("NODEREPORT_DIRECTORY");
+  if (directory_name != NULL) {
+    ProcessNodeReportDirectory(directory_name);
+  }
+  const char* verbose_switch = getenv("NODEREPORT_VERBOSE");
+  if (verbose_switch != NULL) {
+	  nodereport_verbose = ProcessNodeReportVerboseSwitch(verbose_switch);
+  }

```

Comment:
You should probably the getenv calls with a secure_getenv function that looks like this:
```c++
// Look up environment variable unless running as setuid root.
inline const char* secure_getenv(const char* key) {
#ifndef _WIN32
  if (getuid() != geteuid() || getgid() != getegid())
    return nullptr;
#endif
  return getenv(key);
}
```
(From node.cc.)